Beef up the allow_in_graph docs

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #127117

We make the following changes:
- most of the time when someone uses allow_in_graph, they actually
  wanted to make a custom op. We add a link to the custom ops landing
  page and explain the differences between allow_in_graph and custom
  ops.
- we warn people against using allow_in_graph footguns and document
  them.

Test Plan:
- tests

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang

zou3519 committed:
Beef up the allow_in_graph docs

We make the following changes:
- most of the time when someone uses allow_in_graph, they actually
  wanted to make a custom op. We add a link to the Custom Operators
  Manual. You might notice it's a gdoc; our strategy for custom ops docs
  is a combination of tutorials (on pytorch/tutorials) and a single gdoc
  (so we can easily make changes)
- we warn people against using allow_in_graph and document some of its
  restrictions: input/output types and lack of ability to handle
  closures.

Test Plan:
- tests

[ghstack-poisoned]

jansel commented on torch/_dynamo/decorators.py:
```diff
@@ -74,8 +74,40 @@ def assume_constant_result(fn):
+    symbolically introspecting the function properly, then one can decorate said function
+    with allow_in_graph to bypass Dynamo.
+
+    There are a number of restrictions on ``fn``: it must only accept inputs/outputs
+    of certain types (e.g. Tensor/int/bool/float/None and lists of the previous), and
+    all Tensors used inside of ``fn`` must be passed directly as inputs to ``fn``.
+    Failure to abide by these restrictions will result in undefined behavior.
```
Let's be more precise about the restrictions.

jansel commented on torch/_dynamo/decorators.py:
```diff
@@ -74,8 +74,40 @@ def assume_constant_result(fn):
+    do not use this API. Instead, please create a custom operator:
+    https://docs.google.com/document/d/1_W62p8WJOQQUzPsJYa7s701JXt0qf2OfLub2sbkHOaU/edit
+
+    The main use case for :func:`allow_in_graph()` is as an escape hatch for the compiler frontend:
+    if you know the function works w.r.t. to the downstream components of the compilation
+    stack (AOTAutograd and Inductor) but there is a Dynamo bug that prevents it from
+    symbolically introspecting the function properly, then one can decorate said function
+    with allow_in_graph to bypass Dynamo.
```
I don't think the usecase here is only to work around bugs.  I'd suggest adding something like:

| Mechanism      | Frontend (dynamo/symbolic_trace) | Backend (AOTAutograd) |
| -------------- | -------------------- | -------------------- |
| no decorator   | trace inside         | trace inside         | 
| allow_in_graph | opaque callable      | trace inside         |
| wrap (for FX) | opaque callable      | trace inside         |
| custom ops     | opaque callable      | opaque callable      |

Which makes the differences more clear.

zou3519 replied:
Sounds good, I'll add a table

zou3519 replied:
Jason: sometimes people want to execute stuff at compile time: allow_in_graph as a constant. e.g. we allow_in_graph einops (implementation is traceable: string parsing)

Jason is neutral on the rename.

zou3519 commented on torch/_dynamo/decorators.py:
```diff
@@ -74,8 +74,40 @@ def assume_constant_result(fn):
+    If you are trying to black-box a Python function for use with torch.compile
+    (aka something similar to torch.fx.wrap),
+    do not use this API. Instead, please create a custom operator:
+    https://docs.google.com/document/d/1_W62p8WJOQQUzPsJYa7s701JXt0qf2OfLub2sbkHOaU/edit
```
short term: Landing page for custom ops (website) that links to existing tutorials, for more information see the gdoc.
longer term: (2.5) actually toss the gdoc into 1+ webpages

zou3519 commented on torch/_dynamo/decorators.py:
```diff
@@ -74,8 +74,40 @@ def assume_constant_result(fn):
+    and instead directly write it to the graph when encountered.
+
+    If you are trying to black-box a Python function for use with torch.compile
+    (aka something similar to torch.fx.wrap),
```
Don't advertise custom ops as torch.fx.wrap, plop in the table, clarify that this is an fx.wrap if you're not using export

zou3519 commented:
More followups:
# (make issue) zero-tensor meta
# mutate args can be optional (and be pessimistic)
# raw pybind: use warning instead of graph break (because the graph break message gets suppresssed)

zou3519 committed:
Update on "Beef up the allow_in_graph docs"


We make the following changes:
- most of the time when someone uses allow_in_graph, they actually
  wanted to make a custom op. We add a link to the custom ops landing
  page and explain the differences between allow_in_graph and custom
  ops.
- we warn people against using allow_in_graph footguns and document
  them.

Test Plan:
- tests

[ghstack-poisoned]

zou3519 committed:
Update on "Beef up the allow_in_graph docs"


We make the following changes:
- most of the time when someone uses allow_in_graph, they actually
  wanted to make a custom op. We add a link to the custom ops landing
  page and explain the differences between allow_in_graph and custom
  ops.
- we warn people against using allow_in_graph footguns and document
  them.

Test Plan:
- tests

[ghstack-poisoned]

zou3519 committed:
Update on "Beef up the allow_in_graph docs"


We make the following changes:
- most of the time when someone uses allow_in_graph, they actually
  wanted to make a custom op. We add a link to the custom ops landing
  page and explain the differences between allow_in_graph and custom
  ops.
- we warn people against using allow_in_graph footguns and document
  them.

Test Plan:
- tests

[ghstack-poisoned]

zou3519 committed:
Update on "Beef up the allow_in_graph docs"


We make the following changes:
- most of the time when someone uses allow_in_graph, they actually
  wanted to make a custom op. We add a link to the custom ops landing
  page and explain the differences between allow_in_graph and custom
  ops.
- we warn people against using allow_in_graph footguns and document
  them.

Test Plan:
- tests

[ghstack-poisoned]

zou3519 committed:
Update on "Beef up the allow_in_graph docs"


We make the following changes:
- most of the time when someone uses allow_in_graph, they actually
  wanted to make a custom op. We add a link to the custom ops landing
  page and explain the differences between allow_in_graph and custom
  ops.
- we warn people against using allow_in_graph footguns and document
  them.

Test Plan:
- tests

[ghstack-poisoned]

zou3519 commented:
This should be ready for another pass through. I updated the PR with jansel's feedback and our offline discussion:
- removed mentions of torch.fx.wrap, because it's currently confusing (allow_in_graph is a torch.fx.wrap for Dynamo, but not for export today (because export today is Dynamo + AOTAutograd)).
- added more clarity on when to use allow_in_graph and when to use custom ops
- the docs page we link to is now a website

albanD reviewed (approved):
Very clear comparison with custom ops. Thanks!

zou3519 committed:
Update on "Beef up the allow_in_graph docs"


We make the following changes:
- most of the time when someone uses allow_in_graph, they actually
  wanted to make a custom op. We add a link to the custom ops landing
  page and explain the differences between allow_in_graph and custom
  ops.
- we warn people against using allow_in_graph footguns and document
  them.

Test Plan:
- tests

[ghstack-poisoned]

zou3519 commented:
@pytorchbot merge

jansel commented:
@pytorchbot merge

zou3519 commented:
@pytorchbot merge -f "mergebot seems stuck"
